nautobot/extras/plugins/__init__.py

```diff
@@ -491,6 +518,106 @@ def register_filter_extensions(filter_extensions, plugin_name):
                 )
 
 
+#
+# Table Extensions
+#
+
+
+class TableExtension:
+    """Template class for extending Tables.
+
+    An app can override the default columns for a table by either:
+    - Extending the original default columns to include custom columns.
+        - add_to_default_columns = ("my_new_column",)
+    - Removing native columns from the default columns.
+        - remove_from_default_columns = ("tenant",)
+    """
+
+    model = None
+    table_columns = {}
+    add_to_default_columns = ()
+    remove_from_default_columns = ()
+
+    @classmethod
+    def alter_queryset(cls, queryset):
+        """Alter the View class QuerySet.
+
+        This is a good place to add `prefetch_related` to the view queryset.
+        example:
+            return queryset.prefetch_related("my_model_set")
+        """
+        return queryset
+
+
+def register_table_extensions(table_extensions, app_name):
+    """Register a list of TableExtension classes."""
+    for table_extension in table_extensions:
+        _validate_is_subclass_of_table_extension(table_extension)
+        _register_table_base_columns(table_extension, app_name)
+        _modify_default_table_columns(table_extension)
+        _alter_table_view_queryset(table_extension, app_name)
+
+
+def _register_table_base_columns(table_extension, app_name):
+    """Register each column in the TableExtension."""
+    from nautobot.core.utils.lookup import get_table_for_model
+
+    table = get_table_for_model(table_extension.model)
+    for name, column in table_extension.table_columns.items():
+        _validate_table_column_name_is_prefixed_with_app_name(name, app_name)
+        try:
+            _add_column_to_table_base_columns(table, name, column)
+        except AttributeError as error:
+            logger.error(*error.args)
+
+
+def _add_column_to_table_base_columns(table, column_name, column):
+    """Attach a column to an existing table."""
+    import django_tables2
+
+    if not isinstance(column, django_tables2.Column):
+        raise TypeError(f"Custom column `{column_name}` is not an instance of django_tables2.Column.")
+
+    if column_name in table.base_columns:
+        raise AttributeError(f"There was a conflict with table column `{column_name}`, the custom column was ignored.")
+
+    table.base_columns[column_name] = column
+
+
+def _alter_table_view_queryset(table_extension, app_name):
+    """Replace the model view queryset with an optimized queryset from the app."""
+    from nautobot.core.utils.lookup import get_view_for_model
+
+    view = get_view_for_model(table_extension.model, view_type="List")
+    view.queryset = table_extension.alter_queryset(view.queryset)
```

Comment:
Hm. Does this work for both old-style `ObjectListView` as well as `NautobotUIViewSet`? I'm assuming so since the example_app is extending one of each - but I'm a bit concerned in the NautobotUIViewSet case since the class-level `queryset` attribute is shared across all sub-views under the ViewSet. I wonder if in that case it **might** be safer to actually wrap/monkey-patch the `view.alter_queryset()` functions to just call `table_extension.alter_queryset()` when specifically handling a list request?

Reply:
I was concerned about the `NautobotUIViewSet`, which is why I added the code to extend the Circuit table.  This is a good point and I'll look into modifying this behavior.

Reply:
Much appreciate your thoroughness! :-)

Reply:
I am still pending on this.

Reply:
I am unsure of a good way to patch this in the UIViewSet that targets just the list view queryset.  I think I would need to patch either the `.as_view` or the `.get_queryset` method, however, I would sill need to identify how the method is being called to know if it is for a list view or not.  I'm not saying it can't be done, but I'm not sure what the effort would be and if it is worth mucking that deep into the internals.

Reply:
Fair enough. I think maybe for now, we can leave it as is with a TODO comment and see if it causes any issues?

Reply:
Good deal - I've added a TODO on this.